src/coreclr/jit/importer.cpp

```diff
@@ -10098,9 +10098,68 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                             return;
                         }
 
-                        op1 = gtNewOperNode(GT_LCLHEAP, TYP_I_IMPL, op2);
-                        // May throw a stack overflow exception. Obviously, we don't want locallocs to be CSE'd.
-                        op1->gtFlags |= (GTF_EXCEPT | GTF_DONT_CSE);
+                        op1 = gtNewLclHeapNode(op2, opcodeOffs);
+
+                        // Do we have a profile for this non-constant localloc?
+                        // if so, emit "size == profiledValue ? LCLHEAP(profiledValue) : LCLHEAP(size)" tree.
+                        //
+                        // We don't need that optimization if SkipInitLocals is set as even non-constant locallocs
+                        // are cheap in that case, we only want to speed up zeroing.
+                        //
+                        if (info.compInitMem && !op2->IsIntegralConst())
+                        {
+                            // Consuming the existing profile (optimizing)
+                            if (opts.IsOptimizedWithProfile())
+                            {
+                                ssize_t  profiledValue = 0;
+                                uint32_t likelihood    = 0;
+                                if (pickProfiledValue(opcodeOffs, &likelihood, &profiledValue) && (likelihood >= 50) &&
+                                    ((uint32_t)profiledValue <= INT_MAX))
+                                {
+                                    assert(FitsIn<int>(profiledValue));
+
+                                    GenTree* sizeNode          = op2;
+                                    GenTree* clonedSizeNode    = impCloneExpr(sizeNode, &sizeNode, CHECK_SPILL_ALL,
+                                                                              nullptr DEBUGARG("spilling sizeNode"));
+                                    GenTree* profiledValueNode = gtNewIconNode(profiledValue, op2->TypeGet());
+                                    GenTree* fallback          = gtNewLclHeapNode(clonedSizeNode, opcodeOffs);
+                                    fallback->gtFlags |= (GTF_EXCEPT | GTF_DONT_CSE);
+
+                                    GenTree* fastpath;
+                                    if (profiledValue == 0)
+                                    {
+                                        // Just nullptr
+                                        fastpath = gtNewIconNode(0, TYP_I_IMPL);
+                                    }
+                                    else
+                                    {
+                                        // NOTE: we don't want to convert the fastpath stackalloc to a local like we
```

Comment:
If this block is executed frequently enough maybe we should convert to a local? You can compare the block's weight to that of the method entry, if the value is close to 1 then consider the conversion.